pkg/auth/token_handler.go

```diff
@@ -0,0 +1,259 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"go.uber.org/zap"
+	authv1 "k8s.io/api/authentication/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"knative.dev/eventing/pkg/apis/feature"
+	"knative.dev/pkg/configmap"
+	configmapinformer "knative.dev/pkg/configmap/informer"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/system"
+)
+
+type OIDCTokenHandler struct {
+	logger               *zap.Logger
+	kubeClient           kubernetes.Interface
+	provider             *oidc.Provider
+	providerSyncInterval time.Duration
+	providerLock         sync.RWMutex
+	configMapWatcher     configmap.Watcher
+	featureStore         *feature.Store
+}
+
+func NewOIDCTokenHandler(ctx context.Context, kubeClient kubernetes.Interface, logger *zap.Logger) *OIDCTokenHandler {
+	// setup config map watcher, to get informed on changes on the features
+	// configmap to resync the provider config
+	configMapWatcher := configmapinformer.NewInformedWatcher(kubeClient, system.Namespace())
+	featureStore := feature.NewStore(logging.FromContext(ctx).Named("feature-config-store"))
+	featureStore.WatchConfigs(configMapWatcher)
+
+	return &OIDCTokenHandler{
+		logger:               logger,
+		kubeClient:           kubeClient,
+		providerSyncInterval: 5 * time.Minute,
+		configMapWatcher:     configMapWatcher,
+		featureStore:         featureStore,
+	}
+}
+
+// Start the periodic updates of the OIDC provider config and (TBD) token cache.
+// This is done periodically, in case the cluster does not have a valid OIDC
+// config from the beginning or when it gets updated over time (e.g. change of
+// the issuer url).
+// This is a non blocking call.
+func (c *OIDCTokenHandler) Start(stopCh <-chan struct{}) {
+	// Add watch to update provider config on config map changes
+	c.configMapWatcher.Watch(feature.FlagsConfigName, func(cm *v1.ConfigMap) {
+		c.updateProviderConfig()
+	})
+
+	// Start config map watcher
+	if err := c.configMapWatcher.Start(stopCh); err != nil {
+		c.logger.Warn("Failed to start ConfigMap watcher", zap.Error(err))
+	}
+
+	// Load provider config initially
+	if err := c.updateProviderConfig(); err != nil {
+		c.logger.Error("could not update OIDC provider config.", zap.Error(err))
+	}
+
+	// Start periodic provider config syncs
+	go c.syncProviderConfigPeriodically(stopCh)
```

Comment:
If the provider is changed in the API server, I think every emitted token becomes invalid, so even our kubeclient token, do we really need this resync or can we just expect people to restart components if they do such disruptive changes?

Reply:
Updated it and load the config now initially. Makes it easier...